src/swift/Sources/TokenGeneration/NavigationTokens.swift

```diff
@@ -0,0 +1,149 @@
+// --------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation. All rights reserved.
+//
+// The MIT License (MIT)
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the ""Software""), to
+// deal in the Software without restriction, including without limitation the
+// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+// sell copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//
+// --------------------------------------------------------------------------
+
+import AST
+import Foundation
+
+extension TokenFile {
+
+    // MARK: Navigation Emitter Methods
+
+    func navigationTokens(from declarations: [TopLevelDeclaration]) {
+        let packageNavItem = NavigationItem(text: packageName, navigationId: nil, typeKind: .assembly)
+        declarations.forEach { decl in
+            let item = navigationTokens(from: decl)
+            packageNavItem.childItems += item
+        }
+        navigation = [packageNavItem]
+    }
+
+    private func navigationTokens(from decl: TopLevelDeclaration) -> [NavigationItem] {
+        var navItems: [NavigationItem] = []
+        decl.statements.forEach { stmt in
+            if let item = navigationTokens(from: stmt) {
+                navItems.append(item)
+            }
+        }
+        return navItems
+    }
+
+    private func navigationTokens(from decl: ClassDeclaration) -> NavigationItem? {
+        guard publicModifiers.contains(decl.accessLevelModifier ?? .internal) else {
+            return nil
+        }
+        let navItem = NavigationItem(text: decl.name.textDescription, navigationId: decl.name.textDescription, typeKind: .class)
+        decl.members.forEach { member in
+            switch member {
+            case let .declaration(decl):
+                if let item = navigationTokens(from: decl) {
+                    navItem.childItems.append(item)
+                }
+            case .compilerControl:
+                return
+            }
```

Comment:
This switch statement appears several times and could be refactored.

Reply:
Good point. I originally just relocated there here and didn't do any refactoring, but I'll refactor this.

Reply:
Actually, this switch statement can't really be refactored because enums cannot conform to protocols and this would require some kind of protocol to guarantee an enum contained certain cases. The best we could do is refactor lines 61-63 into a one-liner, which I can do.